rhp/v4/server.go

```diff
@@ -275,53 +256,49 @@ func (s *Server) handleRPCModifySectors(stream net.Conn) error {
 	}
 
 	fc := state.Revision
-	cost := prices.RPCModifySectorsCost(req.Actions)
-	// validate the payment without modifying the contract
-	if fc.RenterOutput.Value.Cmp(cost) < 0 {
-		return rhp4.NewRPCError(rhp4.ErrorCodePayment, fmt.Sprintf("renter output value %v is less than cost %v", fc.RenterOutput.Value, cost))
+
+	settings := s.settings.RHP4Settings()
+	if err := req.Validate(s.hostKey.PublicKey(), fc, settings.MaxSectorBatchSize); err != nil {
+		return errorBadRequest("request invalid: %v", err)
 	}
+	prices := req.Prices
 
-	roots := state.Roots
-	for _, action := range req.Actions {
-		switch action.Type {
-		case rhp4.ActionTrim:
-			if action.N > uint64(len(roots)) {
-				return errorBadRequest("trim count %v exceeds sector count %v", action.N, len(roots))
-			}
-			roots = roots[:len(roots)-int(action.N)]
-		case rhp4.ActionUpdate:
-			if action.N >= uint64(len(roots)) {
-				return errorBadRequest("update index %v exceeds sector count %v", action.A, len(roots))
-			}
-			roots[action.N] = action.Root
-		case rhp4.ActionSwap:
-			if action.A >= uint64(len(roots)) || action.B >= uint64(len(roots)) {
-				return errorBadRequest("swap indices %v and %v exceed sector count %v", action.A, action.B, len(roots))
-			}
-			roots[action.A], roots[action.B] = roots[action.B], roots[action.A]
-		default:
-			return errorBadRequest("unknown action type %v", action.Type)
+	// validate that all indices are within the expected range
+	indices := make(map[uint64]bool)
+	for _, i := range req.Indices {
+		if i >= uint64(len(state.Roots)) {
+			return errorBadRequest("index %v exceeds sector count %v", i, len(state.Roots))
 		}
+		indices[i] = true
 	}
 
-	treeHashes, leafHashes := rhp4.BuildModifySectorsProof(req.Actions, state.Roots)
-	resp := rhp4.RPCModifySectorsResponse{
-		OldSubtreeHashes: treeHashes,
-		OldLeafHashes:    leafHashes,
-		NewMerkleRoot:    rhp4.MetaRoot(roots),
+	roots := state.Roots[:0]
+	for i, root := range state.Roots {
+		if indices[uint64(i)] {
+			continue
+		}
+		roots = append(roots, root)
+	}
+
+	// treeHashes, leafHashes := rhp4.BuildModifySectorsProof(req.Actions, state.Roots) // TODO: build proof
```

Comment:
Are the failing tests related to this?

Reply:
The failing test is in`RPCReadSector`

Fail is here: https://github.com/SiaFoundation/coreutils/blob/ba0a3a3731dbd71e37d946fba1fd7d8fae15afc0/rhp/v4/rpc_test.go#L811

Failing logic here: https://github.com/SiaFoundation/coreutils/blob/ba0a3a3731dbd71e37d946fba1fd7d8fae15afc0/rhp/v4/rpc.go#L228

Reply:
Since they're failing on the rhp4 branch as well, I think we can ignore them until the proofs are fixed.